ledger-core/v2/virtual/execute/execute.go

```diff
@@ -0,0 +1,281 @@
+	"github.com/insolar/assured-ledger/ledger-core/v2/insolar"
+	"github.com/insolar/assured-ledger/ledger-core/v2/insolar/payload"
+	"github.com/insolar/assured-ledger/ledger-core/v2/network/messagesender"
+	messageSenderAdapter "github.com/insolar/assured-ledger/ledger-core/v2/network/messagesender/adapter"
+	"github.com/insolar/assured-ledger/ledger-core/v2/reference"
+	"github.com/insolar/assured-ledger/ledger-core/v2/runner"
+	runnerAdapter "github.com/insolar/assured-ledger/ledger-core/v2/runner/adapter"
+	"github.com/insolar/assured-ledger/ledger-core/v2/runner/execution"
+	"github.com/insolar/assured-ledger/ledger-core/v2/runner/executionupdate"
+	"github.com/insolar/assured-ledger/ledger-core/v2/vanilla/injector"
+	"github.com/insolar/assured-ledger/ledger-core/v2/vanilla/throw"
+	"github.com/insolar/assured-ledger/ledger-core/v2/virtual/object"
+)
+
+/* -------- Utilities ------------- */
+
+type SMExecute struct {
+	// input arguments
+	Meta    *payload.Meta
+	Payload *payload.VCallRequest
+
+	// internal data
+	isConstructor      bool
+	isOrdered          bool
+	semaphoreOrdered   smachine.SyncLink
+	semaphoreUnordered smachine.SyncLink
+	execution          execution.Execution
+	objectSharedState  object.SharedStateAccessor
+
+	// execution step
+	executionNewState *executionupdate.ContractExecutionStateUpdate
+	executionID       uuid.UUID
+	executionError    error
+
+	// dependencies
+	runner        *runnerAdapter.Runner
+	messageSender *messageSenderAdapter.MessageSender
+	pulseSlot     *conveyor.PulseSlot
+}
+
+/* -------- Declaration ------------- */
+
+var dSMExecuteInstance smachine.StateMachineDeclaration = &dSMExecute{}
+
+type dSMExecute struct {
+	smachine.StateMachineDeclTemplate
+}
+
+func (*dSMExecute) InjectDependencies(sm smachine.StateMachine, _ smachine.SlotLink, injector *injector.DependencyInjector) {
+	s := sm.(*SMExecute)
+
+	injector.MustInject(&s.runner)
+	injector.MustInject(&s.pulseSlot)
+	injector.MustInject(&s.messageSender)
+}
+
+func (*dSMExecute) GetInitStateFor(sm smachine.StateMachine) smachine.InitFunc {
+	s := sm.(*SMExecute)
+	return s.Init
+}
+
+/* -------- Instance ------------- */
+
+func (s *SMExecute) GetStateMachineDeclaration() smachine.StateMachineDeclaration {
+	return dSMExecuteInstance
+}
+
+func (s *SMExecute) Init(ctx smachine.InitializationContext) smachine.StateUpdate {
+	return ctx.Jump(s.stepWaitObjectReady)
+}
+
+func (s *SMExecute) stepWaitObjectReady(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	var (
+		objectCatalog   = object.Catalog{}
+		callType        = s.Payload.CallType
+		objectID        = s.Payload.Callee.GetLocal()
+		objectReference = reference.NewGlobalSelf(*objectID)
+
+		isOrdered         bool
+		isConstructor     bool
+		objectSharedState object.SharedStateAccessor
+	)
+
+	switch callType {
+	case payload.CTConstructor:
+		objectSharedState = objectCatalog.GetOrCreate(ctx, objectReference)
```

Comment:
we discuss it with Ruz in context of CallConstructor and we realized that it's incorrect, you need to calculate ref of new object